File: python/Trees/same_tree.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        """
        Check if two binary trees are identical.
        Two binary trees are considered the same if they are structurally identical,
        and the nodes have the same value.
        
        Time Complexity: O(min(n,m)) where n and m are the number of nodes in trees p and q
        Space Complexity: O(min(h1,h2)) where h1 and h2 are the heights of trees p and q
        
        Args:
            p (Optional[TreeNode]): Root node of the first binary tree
            q (Optional[TreeNode]): Root node of the second binary tree
            
        Returns:
            bool: True if the trees are identical, False otherwise
        """
        # Base case: if both trees are empty, they are identical
        if not p and not q:
            return True
        # If one tree is empty and the other is not, they are not identical
        # Also check if current nodes have different values
        if not p or not q or p.val != q.val:
            return False
        # Recursively check if left and right subtrees are identical
        return (self.isSameTree(p.left, q.left) and 
                self.isSameTree(p.right, q.right))

# -------------------
# Helper Functions for Testing
# -------------------
def build_tree_from_list(lst):
    if not lst:
        return None
    nodes = [TreeNode(val) if val is not None else None for val in lst]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root

# -------------------
# Test Cases
# -------------------
def test_is_same_tree():
    s = Solution()
    # Test case 1: Identical trees
    p = build_tree_from_list([1,2,3])
    q = build_tree_from_list([1,2,3])
    assert s.isSameTree(p, q) == True
    # Test case 2: Different structure
    p = build_tree_from_list([1,2])
    q = build_tree_from_list([1,None,2])
    assert s.isSameTree(p, q) == False
    # Test case 3: Different values
    p = build_tree_from_list([1,2,1])
    q = build_tree_from_list([1,1,2])
    assert s.isSameTree(p, q) == False
    # Test case 4: Both empty
    p = build_tree_from_list([])
    q = build_tree_from_list([])
    assert s.isSameTree(p, q) == True
    # Test case 5: One empty, one not
    p = build_tree_from_list([1])
    q = build_tree_from_list([])
    assert s.isSameTree(p, q) == False
    print("All test cases passed!")

if __name__ == "__main__":
    test_is_same_tree()

File: C++/Trees/same_tree.cpp
#include <bits/stdc++.h> 
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    /**
     * Check if two binary trees are identical.
     *
     * Two binary trees are considered the same if:
     * - they are structurally identical
     * - their corresponding nodes have the same value
     *
     * Time Complexity: O(min(n, m)) where n and m are number of nodes in the two trees.
     * Space Complexity: O(min(h1, h2)) where h1 and h2 are the heights of the two trees (due to recursion).
     */
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // Base case: if both trees are empty, they are identical
        if (!p && !q) return true;

        // If one tree is empty or values differ, they are not identical
        if (!p || !q || p->val != q->val) return false;

        // Recursively check left and right subtrees
        return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers (iostream, vector, etc.) in one line.

using namespace std;
    ✅ Allows us to write names like vector, cout without the std:: prefix.

struct TreeNode {
    ✅ Defines the binary tree node structure.

int val;
    ✅ Stores the value of the node.

TreeNode *left, *right;
    ✅ Pointers to the left and right children of the node.

TreeNode() : val(0), left(nullptr), right(nullptr) {}
    ✅ Default constructor: sets value = 0 and children = nullptr.

TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    ✅ Constructor with value x and null children.

TreeNode(int x, TreeNode *left, TreeNode *right)
    ✅ Constructor with value x and specified left & right children.

class Solution {
    ✅ Defines the Solution class.

public:
    ✅ Public methods of the Solution class.

bool isSameTree(TreeNode* p, TreeNode* q)
    ✅ Function to check if trees rooted at p and q are identical.

if (!p && !q) return true;
    ✅ If both nodes are null, the trees are identical here.

if (!p || !q || p->val != q->val) return false;
    ✅ If one is null, or their values differ, trees are not identical.

return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    ✅ Recursively check left and right subtrees.

*/
